rlw.6
.TH rlw 6 2022\-09\-05
.SH NAME
rlw - Linux Wrapper for ROBLOX using Wine (Roblox Linux Wrapper)
.SH SYNOPSIS
\fBrlw\fR \fIOPTION\fR
.SH DESCRIPTION
.PP
Roblox Linux Wrapper (\fBrlw\fR) is a Linux Wrapper for ROBLOX using Wine.
It does not implement any command line options other than \fB\-\-help/\-h\fR
and \fB\-\-version/\-v\fR, but rather a graphical selection screen with
various options on start.
.TP
\fB\-\-help/\-h\fR
display this help and exit
.TP
\fB\-\-version/\-v\fR
output version information and exit
.SH "HOMEPAGE"
https://github.com/roblox-linux-wrapper/roblox-linux-wrapper
.SH "SEE ALSO"
WINE(1)


debian/changelog
roblox-linux-wrapper (2017.06.21) unstable; urgency=medium

  * New repository snapshot.
    - Roblox is now started through a plugin-free protocol handler, meaning
      Firefox for Windows is no longer needed.
    - Removed the Legacy Mode launcher (which no longer works) and other
      redundant code.
  * Packaging updates:
    - Move internal code to /usr/share as it is architecture-independent.
    - Bump Standards-version to 4.0.0, no changes needed.
    - Set section to contrib/games.
    - Update my email.

 -- James Lu <bitflip3@gmail.com>  Wed, 21 Jun 2017 20:51:54 -0700

roblox-linux-wrapper (2015.08.13) unstable; urgency=medium

  * Build Git branch master revision 2015.08.13.
  * rlw:
    - export $WINE and $WINESERVER for Winetricks (GH-109).
    - Add handling for --help and --version.
    - Support Debian's wine-development packaging of Wine 1.7.x.

 -- James Lu <glolol1@hotmail.com>  Thu, 13 Aug 2015 02:58:25 -0700

roblox-linux-wrapper (2015.06.16) unstable; urgency=medium

  * Build Git branch master revision 2015.06.16.

 -- James Lu <glolol1@hotmail.com>  Tue, 16 Jun 2015 17:19:30 -0700

roblox-linux-wrapper (2015.05.28) unstable; urgency=medium

  * Build Git branch master revision 2015.05.28.

 -- James Lu <glolol1@hotmail.com>  Thu, 28 May 2015 10:48:38 -0700

roblox-linux-wrapper (2015.05.11+25+gd154d23) unstable; urgency=medium

  * Build Git branch master revision 2015.05.11+25+gd154d23.

 -- James Lu <glolol1@hotmail.com>  Wed, 20 May 2015 11:33:14 -0700

roblox-linux-wrapper (2015.05.11) unstable; urgency=medium

  * Automatic build.

 -- James Lu <glolol1@hotmail.com>  Mon, 11 May 2015 18:15:07 -0700

roblox-linux-wrapper (2015.04.24) unstable; urgency=medium

  * Automatic build.

 -- James Lu <glolol1@hotmail.com>  Fri, 24 Apr 2015 17:25:24 -0700

roblox-linux-wrapper (2015.04.16+10+g617678a) unstable; urgency=medium

  * Initial Release.

 -- James Lu <glolol1@hotmail.com>  Thu, 23 Apr 2015 21:25:13 -0700


RLW.py
# -*- coding: utf-8 -*-
# !/usr/bin/python
"""
Project: roblox-linux-wrapper
File: RLW
Author: Ian
Creation Date: 11/27/2014
"""
from __future__ import print_function
import os
import errno
import subprocess
import sys
import time
import platform
import shutil

try:
    # noinspection PyUnresolvedReferences
    import requests
except ImportError:
    raise noRequests

from which import which
from Errors import NoWine, Unsupported, DownloadError, FatalError, OutdatedPython, noRequests


if platform.system() == "Windows":
    raise Unsupported

if sys.version_info < (3, 0):
    # noinspection PyShadowingBuiltins
    raw_input = input

if not (sys.version_info >= (2, 7)):
    raise OutdatedPython

if (2, 7) <= sys.version_info <= (3, 0):
    pass

__author__ = 'Ian'
__version__ = "20141127b"

WINE = which("wine")
WINESERVER = which("wineserver")
# Uncomment these lines to use Wine Compholio
# WINE = "/opt/wine-compholio/bin/wine"
# WINESERVER = "/opt/wine-compholio/wineserver"

RLWVERSION = __version__
RLWCHANNEL = "PRERELEASE"
WINEPREFIX = os.getenv("HOME") + "/.local/share/wineprefixes/Roblox"
WINEARCH = "win32"
WINETRICKSDEV = "/tmp/winetricks"
WINEDLLOVERRIDES = "winebrowser.exe,winemenubuilder.exe="


def callWine(*args):
    """
    Simplify Calling Wine

    :param args: Arguments
    """
    arg = [WINE]
    for a in args:
        arg.append(a)
    with open(os.devnull, "w") as fnull:
        subprocess.call(arg, stdout = fnull, stderr = fnull)


def srm(filename):
    """
    Silently Remove a file

    :param filename: File to remove
    """
    try:
        os.remove(filename)
    except OSError as e:
        if e.errno != errno.ENOENT:
            raise


# noinspection PyUnresolvedReferences,PyShadowingBuiltins
def wget(file, out):
    """
    Download a file

    :param file: File to download
    :param out: Output Location
    """
    r = requests.get(file, stream = True)

    if not r.ok:
        raise DownloadError

    with open(out, 'wb') as fd:
        start = time.clock()
        total_length = int(r.headers.get('content-length'))
        dl = 0
        if total_length is None:
            fd.write(r.content)
        else:
            for chunk in r.iter_content(1024):
                dl += len(chunk)
                fd.write(chunk)
                done = int((50 * dl) / total_length)
                sys.stdout.write(
                    "\r[%s%s] %s / %s" % ('=' * done, ' ' * (50 - done), (dl // 1024), (total_length // 1024)))
                sys.stdout.flush()
    print("\n")
    print(time.clock() - start)


def checkDeps(Force = False):
    """
    Check Dependencies

    :param Force: Whether to force Dependency download
    """
    print("Checking for required dependencies")
    if not WINE:
        raise NoWine
    if not WINESERVER:
        raise NoWine
    os.environ["WINEPREFIX"] = WINEPREFIX
    os.environ["WINEARCH"] = WINEARCH
    os.environ["WINEDLLOVERRIDES"] = WINEDLLOVERRIDES
    os.environ["WINE"] = WINE
    os.environ["WINESERVERBIN"] = WINESERVER
    os.environ["WINETRICKSDEV"] = WINETRICKSDEV

    if os.path.isfile(WINETRICKSDEV) and os.path.isfile("/tmp/RobloxPlayerLauncher.exe") and \
            os.path.isfile("/tmp/Firefox-Setup-esr.exe") and not Force:
        return
    print("""
    Required dependencies are going to be installed.

    Depending on your internet connection, this may take a few minutes.
    """)

    srm(WINETRICKSDEV)
    wget("http://winetricks.googlecode.com/svn/trunk/src/winetricks", out = "/tmp/winetricks")
    os.chmod('/tmp/winetricks', os.stat('/tmp/winetricks').st_mode | 0o0111)

    srm("/tmp/RobloxPlayerLauncher.exe")
    wget("http://roblox.com/install/setup.ashx", out = "/tmp/RobloxPlayerLauncher.exe")

    srm("/tmp/Firefox-Setup-esr.exe")
    wget(
        "http://ftp.mozilla.org/pub/mozilla.org/firefox/releases/31.1.1esr/win32/en-US/Firefox%20Setup%2031.1.1esr.exe",
        out = "/tmp/Firefox-Setup-esr.exe")

    Install()


def Install():
    """
    Install RLW

    """
    print("""
    Installing Dependencies, please wait.
    """)
    with open(os.devnull, "w") as fnull:
        subprocess.call(["/tmp/winetricks",
                         "-q",
                         "ddr=gdi",
                         "vcrun2008",
                         # "mshtml",
                         # "mshttp",
                         "vcrun2012",
                         "vcrun2013",
                         "winhttp",
                         "wininet",
                         "wmp9",
                        ], stdout = fnull, stderr = fnull)

        callWine("/tmp/RobloxPlayerLauncher.exe")

        ROBLOXPROXY = None

        rootdir = WINEPREFIX
        for dirName, subdirList, fileList in os.walk(rootdir):
            if "RobloxProxy.dll" in fileList:
                ROBLOXPROXY = dirName + "/RobloxProxy.dll"

        if ROBLOXPROXY is None:
            raise FatalError

        callWine("regsvr32", "/i", ROBLOXPROXY)

        callWine("/tmp/Firefox-Setup-esr.exe", "/SD")


def main():
    """
    Main Program for RLW

    """
    print('Roblox Linux Wrapper v' + RLWVERSION + '-' + RLWCHANNEL)
    installed = os.path.isdir(os.getenv("HOME") + "/.rlw")
    print("""
          1. Play Roblox
          2. Play Roblox (Legacy)
          3. Roblox Studio
          4. {0}
          5. Reset Roblox to defaults/Reinstall ROBLOX
          6. Uninstall Roblox
          7. Exit""".format(
        installed and "Uninstall Roblox Linux Wrapper" or "Install Roblox Linux Wrapper (Recommended)")
    )
    choice = raw_input("> ")
    if choice == 1:
        callWine("C:\Program Files\Mozilla Firefox\\firefox.exe", "http://www.roblox.com/Games.aspx")
    if choice == 2:
        PLAYER = None
        for dirName, subdirList, fileList in os.walk(WINEPREFIX):
            if "RobloxPlayerBeta.exe" in fileList:
                PLAYER = dirName + "/RobloxPlayerBeta.exe"
        x = raw_input("GameID: ")
        callWine(PLAYER, "--id", str(x))
    if choice == 3:
        for dirName, subdirList, fileList in os.walk(WINEPREFIX):
            if "RobloxStudioLauncherBeta.exe" in fileList:
                callWine(dirName + "/RobloxStudioLauncherBeta.exe", "-ide")
                subprocess.call([WINESERVER, "-k"])
        for dirName, subdirList, fileList in os.walk(WINEPREFIX):
            if "RobloxStudioBeta.exe" in fileList:
                callWine(dirName + "/RobloxStudioBeta.exe")
    if choice == 4 and not installed:
        di = os.getenv("HOME") + "/.local/share/applications/"
        with open(di + "Roblox.desktop", "w") as f:
            f.write("""
            [Desktop Entry]
            Comment=Play Roblox
            Name=Roblox Linux Wrapper
            Exec=$HOME/.rlw/RLW.py
            Actions=RFAGroup;ROLWiki;
            GenericName=Building Game
            Icon=roblox
            Categories=Game;
            Type=Application

            [Desktop Action ROLWiki]
            Name='Roblox on Linux Wiki'
            Exec=xdg-open 'http://roblox.wikia.com/wiki/Roblox_On_Linux'

            [Desktop Action RFAGroup]
            Name='Roblox for All'
            Exec=xdg-open 'http://www.roblox.com/Groups/group.aspx?gid=292611'""")
            f.close()
        try:
            os.mkdir(os.getenv("HOME") + "/.rlw")
            shutil.copy(__file__, os.getenv("HOME") + "/.rlw/RLW.py")
        except OSError as e:
            if e.errno == 17:
                pass
        # TODO: Download link to latest version of this python script, once uploaded.
        wget("http://img1.wikia.nocookie.net/__cb20130302012343/robloxhelp/images/f/fb/ROBLOX_Circle_Logo.png",
             out = os.getenv("HOME") + "/.local/share/icons/roblox.png")
        os.chmod(os.getenv("HOME") + "/.rlw", os.stat(os.getenv("HOME") + "/.rlw").st_mode | 0o0111)
        os.chmod(os.getenv("HOME") + "/.local/share/applications/Roblox.desktop",
                 os.stat(os.getenv("HOME") + "/.local/share/applications/Roblox.desktop").st_mode | 0o0111)
        subprocess.call(["xdg-desktop-menu",
                         "install",
                         "--novendor",
                         os.getenv("HOME") + "/.local/share/applications/Roblox.desktop"
        ])
        subprocess.call(["xdg-desktop-menu", "forceupdate"])
        print("Roblox Linux Wrapper was installed successfully.")
    if choice == 4 and installed:
        subprocess.call(["xdg-desktop-menu",
                         "uninstall",
                         os.getenv("HOME") + "/.local/share/applications/Roblox.desktop"
        ])
        shutil.rmtree(os.getenv("HOME") + "/.rlw")
        os.remove(os.getenv("HOME") + "/.local/share/icons/roblox.png")
        subprocess.call(["xdg-desktop-menu", "forceupdate"])
        print("Roblox Linux Wrapper has been uninstalled successfully.")
    if choice == 5:
        try:
            shutil.rmtree(WINEPREFIX)
        except OSError as e:
            if e.errno == 2:
                pass
        checkDeps(Force = True)
        main()
    if choice == 6:
        subprocess.call([WINESERVER, "-k"])
        try:
            shutil.rmtree(WINEPREFIX)
        except OSError as e:
            if e.errno == 2:
                pass
        print("Uninstalled ROBLOX")
    if choice == 7:
        sys.exit()
    os.system("clear")
    main()


try:
    os.system("clear")
    checkDeps()
    main()
except KeyboardInterrupt:
    print("\r")
    print("Exiting RLW")
    pass
